Sources/Commands/SwiftRunTool.swift

```diff
@@ -0,0 +1,148 @@
+    case noExecutableFound
+    case executableNotFound(String)
+    case multipleExecutables
+}
+
+extension RunError: CustomStringConvertible {
+    var description: String {
+        switch self {
+        case .noExecutableFound:
+            return "no executable product found"
+        case .executableNotFound(let executable):
+            return "could not find executable '\(executable)'"
+        case .multipleExecutables:
+            return "multiple products defined"
+        }
+    }
+}
+
+public class RunToolOptions: ToolOptions {
+    /// Returns the mode in with the tool command should run.
+    var mode: RunMode {
+        // If we got version option, just print the version and exit.
+        if shouldPrintVersion {
+            return .version
+        }
+
+        return .run
+    }
+    
+    /// If the executable product should be built before running.
+    var shouldBuild = true
+    
+    /// The custom working directory the executable should be run in.
+    var workingDirectory: AbsolutePath?
+    
+    /// The executable product to run.
+    var executable: String?
+    
+    /// The arguments to pass to the executable.
+    var arguments: [String] = []
+}
+
+public enum RunMode {
+    case version
+    case run
+}
+
+/// swift-run tool namespace
+public class SwiftRunTool: SwiftTool<RunToolOptions> {
+
+   public convenience init(args: [String]) {
+       self.init(
+            toolName: "run",
+            usage: "[options] [executable <arguments>]",
+            overview: "Build and run an executable product",
+            args: args
+        )
+    }
+
+    override func runImpl() throws {
+        switch options.mode {
+        case .version:
+            print(Versioning.currentVersion.completeDisplayString)
+
+        case .run:
+            let plan = try buildPlan()
+
+            if options.shouldBuild {
+                try build(plan: plan, includingTests: false)
+            }
+
+            let product = try findExecutable(in: plan)
+            let path = plan.buildParameters.buildPath.appending(component: product.name)
+            try run(at: path)
+        }
+    }
+
+    /// Find executable product based on options.
+    private func findExecutable(in plan: BuildPlan) throws -> ResolvedProduct {
+        let executableProducts = plan.graph.products.filter({ $0.type == .executable })
+        
+        guard executableProducts.count > 0 else {
+            throw RunError.noExecutableFound
+        }
+        
+        if let executable = options.executable {
+            guard let executableProduct = executableProducts.first(where: { $0.name == executable }) else {
+                throw RunError.executableNotFound(executable)
+            }
+            
+            return executableProduct
+        } else {
+            guard executableProducts.count == 1 else {
+                throw RunError.multipleExecutables
+            }
+            
+            return executableProducts.first!
```

Comment:
Ya, I wouldn't block this one on it, just update it to use it if we decide it is a good idea.